pkg/deviceclaimingserver/enddevices/enddevices.go

```diff
@@ -179,15 +178,19 @@ func (upstream *Upstream) Claim(ctx context.Context, req *ttnpb.ClaimEndDeviceRe
 	if authenticatedIDs := req.GetAuthenticatedIdentifiers(); authenticatedIDs != nil {
 		joinEUI = req.GetAuthenticatedIdentifiers().JoinEui
 	} else if qrCode := req.GetQrCode(); qrCode != nil {
-		data, err := qrcode.Parse(qrCode)
+		conn, err := upstream.Component.GetPeerConn(ctx, ttnpb.ClusterRole_QR_CODE_GENERATOR, nil)
 		if err != nil {
-			return nil, errParseQRCode.WithCause(err)
+			return nil, err
 		}
-		authIDs, ok := data.(qrcode.AuthenticatedEndDeviceIdentifiers)
-		if !ok {
-			return nil, errQRCodeData.New()
+		qrg := ttnpb.NewEndDeviceQRCodeGeneratorClient(conn)
+		data, err := qrg.Parse(ctx, &ttnpb.ParseEndDeviceQRCodeRequest{
+			QrCode: qrCode,
+		})
```

Comment:
So I still would like to see some form of authentication here, even if it was forwarded auth or cluster auth or whatever, so we can do proper rate limiting on the authenticated party.

Reply:
Ok then I think that the same auth check must be applied for the other RPCs in the service for the same reasons. I've done that as well.